group/assignment5/Assignment 5.java

```diff
@@ -0,0 +1,257 @@
+            char newChar = curWord.charAt(i1);
+            if (!visitedChars.containsKey(newChar)) {
+                Node newNode = new Node(newChar);
+                root.addChild(newNode);
+                newNode.incNumOfParents();
+                visitedChars.put(newChar, newNode);
+            }
+        }
+    }
+
+    private static Node makeGraph(List<String> dictionary) {
+
+        //base node
+        Node root = new Node(' ');
+
+        //using hashMap so that we have only one node for one char
+        HashMap<Character, Node> visitedChars = new HashMap<Character, Node>();
+
+        String prevWord;
+        String curWord;
+
+        
+        prevWord = dictionary.get(0);
+        int i = 1;
+        while (i < dictionary.size()) {
+            curWord = dictionary.get(i);
+            compareWords(prevWord, curWord, root, visitedChars);
+            prevWord = curWord;
+            i++;
+        }
+
+        return root;
+    }
+
+    private static Node findNextNode(HashSet<Node> leftNodes) {
+
+        //looking for a node without any parents
+        Iterator<Node> iterator = leftNodes.iterator();
+        while (iterator.hasNext()) {
+            Node next = iterator.next();
+            if (next.getNumOfParents() == 0) {
+                return next;
+            }
+        }
+
+        return null;
+    }
+
+    private static List<Character> topSort(Node root) {
+        List<Character> charOrder = new LinkedList<>();
+        HashSet<Node> leftNodes = new HashSet<Node>();
+        List<Node> rootsChildren = root.getChildren();
+        for (Node n : rootsChildren) {
+            leftNodes.add(n);
+            n.decNumOfParents();
+        }
+        root.removeChildren();
+        Node nextNode = findNextNode(leftNodes);
+        while (nextNode != null) {
+            //delete found node from the waiting list
+            leftNodes.remove(nextNode);
+            //delete dependencies from the found node to his children
+            List<Node> nextNodesChildren = nextNode.getChildren();
+            for (Node n : nextNodesChildren) {
+                leftNodes.add(n);
+                n.decNumOfParents();
+            }
+            nextNode.removeChildren();
+            charOrder.add(nextNode.getValue());
+            nextNode = findNextNode(leftNodes);
+        }
+        if (leftNodes.size() > 0) {
+            System.out.println("Cycle!");
+            return null; //there is a cycle, could throw an exception
+        }
+        return charOrder;
+    }
+
+    public static List<Character> getAlphabet(List<String> dictionary) {
+        if (dictionary == null) {
+            return null;
+        }
+        List<Character> charOrder = null;
+        if (dictionary.size() <= 1) {
+            charOrder = buildOneWordSolution(dictionary.get(0));
+            return charOrder;
+        }
+
+        Node root = makeGraph(dictionary);
+        charOrder = topSort(root);
+        return charOrder;
+    }
+

```

Comment:
I'd say a separate new file or a junit test file for tests would be better, for avoiding too much functionality in one file.